cips/cip-31.md

```diff
@@ -0,0 +1,157 @@
+
+3. **Token State Management:**  
+   Tokens are tracked in four states:  
+
+   - Locked & Unstaked  
+   - Locked & Staked  
+   - Unlocked & Unstaked  
+   - Unlocked & Staked
+
+4. **Mathematical Model & Example:**  
+   For a lockup schedule of 100 tokens over 365 days (≈0.274 tokens/day), once 30% of the tokens have unlocked, claiming a reward of about 0.657 tokens immediately unlocks 30% (≈0.197 tokens) while the remaining 70% (≈0.460 tokens) is added to the locked balance. This brings the locked amount to roughly 70.460 tokens, which will continue unlocking over the remaining 70% of the period (approximately 255.5 days), yielding an updated daily unlock rate of about 0.276 tokens/day.
+
+    *NOTE: reward calculation is arbitrary, it is not based on Celestia's inflation.*
+
+5. **Error Handling:**  
+   - Ensure sufficient balance checks before processing rewards.  
+   - Validate that the lockup period is still active.  
+   - Separate commission transfers from reward lockup updates.
+
+### Module Interaction
+
+The following sequence summarizes the module interactions:
+
+```mermaid
+sequenceDiagram
+participant DM as Distribution Module
+participant AK as Account Keeper
+participant LM as Lockup Module
+participant BM as Bank Module
+
+DM->>AK: Get Receiving Account
+AK-->>DM: Return Account
+
+rect rgb(0, 0, 139)
+  Note over DM: Account Type Check
+  DM->>DM: Is Lockup Account?
+end
+
+alt Is Lockup Account
+   DM->>LM: Get Lockup Type
+   DM->>LM: Calculate New Lockup Schedule
+   LM->>LM: Update Lockup Parameters
+   LM->>BM: Update Account Balance
+   LM-->>DM: Return Success
+else Not Lockup Account
+  DM->>BM: Process Normal Reward Transfer
+end
+
+BM-->>DM: Return Status
+
+
+```
+
+### Commission Cap
+
+As part of this CIP we also propose capping the validator commission at 25%. A commission cap is required to avoid circumvention of this CIP by large token holders, who could spin up a new validator with 100% commission, where such commission would be fully unlocked. 
+
+The commission cap will also be applied to validators who were created prior to the upgrade but are unbonded. This prevents new validators from being created today in order to avoid the cap in the future.  
+
+# Parameters
+
+When upgrading to v4 we propose introducing a migration that will set the Parameter in [`x/Staking`](https://github.com/cosmos/cosmos-sdk/blob/release/v0.50.x/x/staking/types/staking.pb.go#L934) to 25% disallowing new validators from creating validators with \>25% commission rates. 
+
+# Backwards Compatibility
+
+No backwards incompatibilities are introduced for non-lockup accounts. Existing lockup accounts not subject to the new reward integration will continue to operate as before. For accounts that do use the new functionality, the changes are additive and maintain the original lockup semantics.
```

Comment:
> Existing lockup accounts not subject to the new reward integration will continue to operate as before.

Which existing lockup accounts are not subject to this CIP? It's not entirely clear to me if this CIP impacts existing lockup accounts or only newly created lockup accounts. I assume the former.